tools/dynamicdatamonitor/DynamicDataMonitor.cpp

```diff
@@ -0,0 +1,181 @@
+/*
+ *
+ *
+ * Distributed under the OpenDDS License.
+ * See: http://www.opendds.org/license.html
+ */
+
+#include <dds/DCPS/transport/framework/TransportRegistry.h>
+#include <dds/DCPS/transport/framework/TransportConfig.h>
+#include <dds/DCPS/transport/framework/TransportInst.h>
+
+#include <dds/DCPS/Recorder.h>
+#include <dds/DdsDcpsInfrastructureC.h>
+#include <dds/DCPS/Marked_Default_Qos.h>
+#include <dds/DCPS/Service_Participant.h>
+#include <dds/DCPS/SubscriberImpl.h>
+#include <dds/DCPS/WaitSet.h>
+#include <dds/DCPS/Serializer.h>
+#include <dds/DCPS/XTypes/TypeObject.h>
+#include <dds/DCPS/XTypes/DynamicData.h>
+#include "dds/DCPS/StaticIncludes.h"
+#if defined ACE_AS_STATIC_LIBS && !defined OPENDDS_SAFETY_PROFILE
+#include <dds/DCPS/transport/udp/Udp.h>
+#include <dds/DCPS/transport/multicast/Multicast.h>
+#include <dds/DCPS/RTPS/RtpsDiscovery.h>
+#include <dds/DCPS/transport/rtps_udp/RtpsUdp.h>
+#include <dds/DCPS/transport/shmem/Shmem.h>
+#endif
+
+#include <ace/Semaphore.h>
+#include <ace/Thread_Semaphore.h>
+
+#include <iostream>
+#include <sstream>
+
+class TestRecorderListener : public OpenDDS::DCPS::RecorderListener
+{
+public:
+  explicit TestRecorderListener()
+    : sem_(0)
+  {
+  }
+
+  virtual void on_sample_data_received(OpenDDS::DCPS::Recorder* rec,
+                                       const OpenDDS::DCPS::RawDataSample& sample)
+  {
+    using namespace OpenDDS::DCPS;
+    OpenDDS::XTypes::DynamicData dd = rec->get_dynamic_data(sample);
+    OpenDDS::DCPS::String my_type = "";
+    if (!print_dynamic_data(dd, dd.get_type(), my_type, "")){
+      ACE_ERROR((LM_ERROR, "Error: (%P|%t) on_sample_data_received Failed to read dynamic data\n"));
+    }
+    std::cout << my_type;
+  }
+
+  virtual void on_recorder_matched(OpenDDS::DCPS::Recorder*,
+                                   const ::DDS::SubscriptionMatchedStatus& status )
+  {
+    if (status.current_count == 1) {
+        ACE_DEBUG((LM_DEBUG, ACE_TEXT("TestRecorderListener -- a writer connect to recorder\n")));
+    }
+    else if (status.current_count == 0 && status.total_count > 0) {
+      ACE_DEBUG((LM_DEBUG, ACE_TEXT("TestRecorderListener -- writer disconnect with recorder\n")));
+      sem_.release();
+    }
+  }
+
+  int wait(const ACE_Time_Value & tv) {
+    ACE_Time_Value timeout = ACE_OS::gettimeofday() + tv;
+    return sem_.acquire(timeout);
+  }
+private:
+  ACE_Thread_Semaphore sem_;
+};
+
+
+int run_test(int argc, ACE_TCHAR *argv[]){
+  int ret_val = 0;
+  try {
+    OpenDDS::DCPS::String topic_name = argv[1];
+    OpenDDS::DCPS::String type_name = argv[2];
+    DDS::DomainId_t domainid = ACE_OS::atoi(argv[3]);
```

Comment:
Please check that there are that many arguments and don't check the arguments until after `TheParticipantFactoryWithArgs`. `TheParticipantFactoryWithArgs` pulls out the `-DCPS*` options.